src/transformers/models/segformer/modeling_tf_segformer.py

```diff
@@ -0,0 +1,881 @@
+                encoder_hidden_state = tf.reshape(encoder_hidden_state, (batch_size, height, width, -1))
+
+            # unify channel dimension
+            encoder_hidden_state = tf.transpose(encoder_hidden_state, perm=[0, 2, 3, 1])
+            height = shape_list(encoder_hidden_state)[1]
+            width = shape_list(encoder_hidden_state)[2]
+            encoder_hidden_state = mlp(encoder_hidden_state)
+            encoder_hidden_state = tf.reshape(encoder_hidden_state, (batch_size, height, width, -1))
+
+            # upsample
+            temp_state = tf.transpose(encoder_hidden_states[0], perm=[0, 2, 3, 1])
+            upsample_resolution = shape_list(temp_state)[1:-1]
+            encoder_hidden_state = tf.image.resize(encoder_hidden_state, size=upsample_resolution, method="bilinear")
+            all_hidden_states += (encoder_hidden_state,)
+
+        hidden_states = self.linear_fuse(tf.concat(all_hidden_states[::-1], axis=-1))
+        hidden_states = self.batch_norm(hidden_states)
+        hidden_states = self.activation(hidden_states)
+        hidden_states = self.dropout(hidden_states)
+
+        # logits of shape (batch_size, height/4, width/4, num_labels)
+        logits = self.classifier(hidden_states)
+
+        return logits
+
+
+@add_start_docstrings(
+    """SegFormer Model transformer with an all-MLP decode head on top e.g. for ADE20k, CityScapes.""",
+    SEGFORMER_START_DOCSTRING,
+)
+class TFSegformerForSemanticSegmentation(TFSegformerPreTrainedModel):
+    def __init__(self, config: SegformerConfig, **kwargs):
+        super().__init__(config, **kwargs)
+        self.segformer = TFSegformerMainLayer(config, name="segformer")
+        self.decode_head = TFSegformerDecodeHead(config, name="decode_head")
+
+    def hf_compute_loss(self, logits, labels):
+        # upsample logits to the images' original size
+        if len(shape_list(labels)) > 3:
+            label_interp_shape = shape_list(labels)[1:-1]
+        else:
+            label_interp_shape = shape_list(labels)[-2:]
+
+        upsampled_logits = tf.image.resize(logits, size=label_interp_shape, method="bilinear")
+        # compute weighted loss
+        loss_fct = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True, reduction="none")
+
+        # Adapted from https://www.tensorflow.org/text/tutorials/transformer#loss_and_metrics.
+        # Utility to mask the index to ignore during computing the loss.
+        def masked_loss(real, pred):
+            mask = tf.math.logical_not(tf.math.equal(real, self.config.semantic_loss_ignore_index))
+            loss_ = loss_fct(real, pred)
+            mask = tf.cast(mask, dtype=loss_.dtype)
+            loss_ *= mask
+            # We return loss per-sample.
+            # Initially the loss is of shape - (batch_size, height, width), then
+            # it is reduced across the spatial resolutions per-sample.
+            return tf.reduce_sum(loss_, axis=(1, 2)) / tf.reduce_sum(mask, axis=(1, 2))
```

Comment:
@Rocketknight1 new loss computation with the reduction scheme we discussed.

Reply:
@sayakpaul This looks good to me! I discussed something similar to this with @ydshieh though - after talking about it this morning, he pointed out that a per-sample loss may cause issues with tf-pt equivalence tests. The reason is that if samples have different numbers of masked positions, then the final loss changes depending on whether you compute a scalar loss directly, or compute a per-sample loss and then take the mean of that later.

This isn't a major issue, though - if tests pass with the per-sample loss here, I think that's preferable. Still, we should keep it in mind in case we get pt-tf test failures!

Reply:
Thanks for explaining, @Rocketknight1! 

I have run both `RUN_SLOW=1 RUN_PT_TF_CROSS_TESTS=1 python -m pytest tests/models/segformer/test_modeling_tf_segformer.py` and `RUN_SLOW=1 RUN_PT_TF_CROSS_TESTS=1 python -m pytest tests/models/segformer/test_modeling_segformer.py` and all the tests passed. But as you mentioned we should keep this nuance about the loss and how it impacts the PT-TF tests in mind going forward.

Reply:
Hello. By reading the loss implementation in `SegformerForSemanticSegmentation`

https://github.com/huggingface/transformers/blob/f0982682bd6fd0b438dda79ec45f3a8fac83a985/src/transformers/models/segformer/modeling_segformer.py#L814-L816

I believe it is based on pixel-value (which is similar to token level in NLP).

The reason that the test passed is probably because `config.semantic_loss_ignore_index` is not used in `(TF)SegformerModelTester` to generate the segmentation labels.

I would advocate to verify this, improve the tests if the above is True, and change the TF loss to the desired one.

cc @Rocketknight1  @NielsRogge

Reply:
This is what I see about labels preparation

In PyTorch
```
            elif model_class in get_values(MODEL_FOR_SEMANTIC_SEGMENTATION_MAPPING):
                batch_size, num_channels, height, width = inputs_dict["pixel_values"].shape
                inputs_dict["labels"] = torch.zeros(
                    [self.model_tester.batch_size, height, width], device=torch_device
                ).long()
```
In TF
```
...
                inputs_dict["labels"] = tf.zeros(
                    (self.model_tester.batch_size, self.model_tester.seq_length), dtype=tf.int32
                )
```

However, we have `semantic_loss_ignore_index=255`.

Reply:
So, what do you suggest?

Reply:
Hi, the code snippet is `SegformerForSemanticSegmentation`, so not sequence classification problems, unless I misunderstand what you mean here.

Reply:
The PT/TF equivalence test for `SegformerForSemanticSegmentation` is likely not run with `labels`

https://github.com/huggingface/transformers/blob/f0982682bd6fd0b438dda79ec45f3a8fac83a985/tests/test_modeling_tf_common.py#L593-L603

as @sayakpaul pointed out we don't prepare labels for semantic segmentation in TF

https://github.com/huggingface/transformers/blob/f0982682bd6fd0b438dda79ec45f3a8fac83a985/tests/test_modeling_tf_common.py#L136

Reply:
The PT/TF equivalence test for `SegformerForSemanticSegmentation` is likely not run with `labels`

https://github.com/huggingface/transformers/blob/f0982682bd6fd0b438dda79ec45f3a8fac83a985/tests/test_modeling_tf_common.py#L593-L603

as @sayakpaul pointed out we don't prepare labels for semantic segmentation in TF

https://github.com/huggingface/transformers/blob/f0982682bd6fd0b438dda79ec45f3a8fac83a985/tests/test_modeling_tf_common.py#L136

Reply:
> I am quite sure that we will have to change the current per-sample loss in TFSegformerForSemanticSegmentation though, and would prefer to have a computation that corresponds to what SegformerForSemanticSegmentation has in this PR.

The PT model returns a scalar i.e., reduced across all the samples as well as the spatial dimensions of the labels (segmentation maps). It's not a problem for me to do that for the TF model and change the overridden `test_loss_computation()` method. 

But here's another consideration: https://github.com/huggingface/transformers/pull/17910#discussion_r912186215.

Reply:
The discussion I had with @Rocketknight1 : https://github.com/huggingface/transformers/pull/17973#issuecomment-1173485211

In short: This depends on the loss is sample-based (per sentence/pixel_values, etc.) or token/pixel-based. In the later case, averaging over non-batch dimensions to get per sample losses is questionable.

Reply:
Cool. So, what changes would you suggest?

Reply:
`tf.reduce_sum(loss_) / tf.reduce_sum(mask)` should be fine.

similar to https://github.com/huggingface/transformers/blob/ec07eccc7d61a77ca3c0463f67bcde18b9943fea/src/transformers/modeling_tf_utils.py#L210

Reply:
Alright doing it. Will also reflect this from the `test_loss_computation()` method.

Reply:
@ydshieh @Rocketknight1 I have incorporated the changes in the loss computation and the corresponding test. PTAL.